java/ql/src/Telemetry/AutomodelSharedCharacteristics.qll

```diff
@@ -0,0 +1,285 @@
+   * DataFlow node class, or a subtype thereof.
+   */
+  class Endpoint;
+
+  /**
+   * A related location for an endpoint. This will typically be bound to the supertype of all AST nodes (eg., `Top`).
+   */
+  class RelatedLocation;
+
+  /**
+   * A class label for an endpoint.
+   */
+  class EndpointType extends string;
+
+  /**
+   * An EndpointType that denotes the absence of any sink.
+   */
+  class NegativeEndpointType extends EndpointType;
+
+  RelatedLocation asLocation(Endpoint e);
+
+  /**
+   * Defines what MaD labels are known, and what endpoint type they correspond to.
+   */
+  predicate isKnownLabel(string label, string humanReadableLabel, EndpointType type);
+
+  /**
+   * Should hold for any endpoint that is a flow sanitizer.
+   */
+  predicate isSanitizer(Endpoint e, EndpointType t);
+
+  /**
+   * Should hold for any endpoint that is a sink of the given (known or unknown) label.
+   */
+  predicate isSink(Endpoint e, string label);
+
+  /**
+   * Should hold for any endpoint that is known to not be any sink.
+   */
+  predicate isNeutral(Endpoint e);
+
+  RelatedLocation getRelatedLocation(Endpoint e, string name);
+}
+
+/**
+ * A set of shared characteristics for a given candidate class.
+ *
+ * This module is language-agnostic, although the `CandidateSig` module will be language-specific.
+ *
+ * The language specific implementation can also further extend the behavior of this module by adding additional
+ *   implementations of endpoint characteristics exported by this module.
+ */
+module SharedCharacteristics<CandidateSig Candidate> {
+  predicate isSink = Candidate::isSink/2;
+
+  predicate isNeutral = Candidate::isNeutral/1;
+
+  /**
+   * Holds if `sink` is a known sink of type `endpointType`.
+   */
+  predicate isKnownSink(Candidate::Endpoint sink, Candidate::EndpointType endpointType) {
+    // If the list of characteristics includes positive indicators with maximal confidence for this class, then it's a
+    // known sink for the class.
+    not endpointType instanceof Candidate::NegativeEndpointType and
+    exists(EndpointCharacteristic characteristic |
+      characteristic.appliesToEndpoint(sink) and
+      characteristic.hasImplications(endpointType, true, maximalConfidence())
+    )
+  }
+
+  /**
+   * Holds if the candidate sink `candidateSink` should be considered as a possible sink of type `sinkType`, and
+   * classified by the ML model. A candidate sink is a node that cannot be excluded from `sinkType` based on its
+   * characteristics.
+   */
+  predicate isSinkCandidate(Candidate::Endpoint candidateSink, Candidate::EndpointType sinkType) {
+    not sinkType instanceof Candidate::NegativeEndpointType and
+    not exists(getAReasonSinkExcluded(candidateSink, sinkType))
+  }
+
+  /**
+   * If it exists, gets a related location for a given endpoint or candidate.
+   * If it doesn't exist, returns the candidate itself as a 'null' value.
```

Comment:
Small NIT + what do you mean by 'null' value? Should we just remove that mention? QLDoc should describe what the predicate does, not what it's going to be used for.

```suggestion
   * Gets the related location of `e` with name `name`, if it exists.
   * Otherwise, gets the candidate itself.
```

Reply:
:+1: accepted the suggestion! (will push later)